### PR TITLE
Prevent generation of redundant file elements

### PR DESCRIPTION
**Pull request by kwridan**
Resolves https://github.com/tuist/tuist/issues/514

### Short description 📝

File elements for all dependencies and products within a graph were always generated and included in projects without applying any filtering.
This led to the inclusion of redundant file elements in top level projects (for products & dependencies that they didn't require)

In this example:

Dependencies:
- Framework2 -> Framework3
- Framework4 -> Framework4
- Framework5 -> `ARKit.framework` however

The highlighted elements are redundant 

Screenshot 2019-09-28 at 16 22 45

### Solution 📦

Leverage the business logic with the `Graph` to determine which file references are needed.

### Implementation 👩‍💻👨‍💻

- [x] Update `DependencyReference.product` to include product name
- [x] Add helper methods to `Graph` to allow listing all dependency references for a project
- [x] Update `ProjectFileElements` to leverage new helper
- [x] Update helpers to also resolve local swift packages
- [x] Add fixture
- [x] Update change log

### Test Plan 🛠:

- Generate the fixture `fixtures/ios_app_with_frameworks` via `tuist generate`
- Open the generated Workspace
- Verify the `Framework3` project doesn't contain file references for `Framework5`
- Verify the `Framework3` project doesn't contain file references for `ARKit.framework`
- Verify the `Framework4` project doesn't contain file references for `ARKit.framework`

### CODE REVIEW

**Comment by pepibumur** [comment]
Is this ready for review @kwridan ?

**Comment by kwridan** [comment]
will rebase it ontop of master, there are a few additional changes needed for the swift packages dependencies too

File: Sources/TuistGenerator/Graph/Graph.swift [hunk 1]
```diff
@@ -305,13 +311,44 @@ class Graph: Graphing {
...
         let otherTargetFrameworks = findAll(targetNode: targetNode, test: isFramework)
-            .map { DependencyReference.product(target: $0.target.name) }
+            .map { DependencyReference.product(target: $0.target.name, productName: $0.target.productNameWithExtension) }
 
         references.append(contentsOf: otherTargetFrameworks)
 
         return Set(references).sorted()
     }
 
+    func copyProductDependencies(path: AbsolutePath, target: Target) -> [DependencyReference] {
```
Comment (pepibumur) [inline Sources/TuistGenerator/Graph/Graph.swift]: I think `copyProductDependencies` is a bit ambiguous. What does it mean? What about including a note like [this one](https://github.com/tuist/tuist/pull/515/files#diff-bd14a50b734b61af028ddef07944c580L279) explaining why this method is useful?

File: Sources/TuistGenerator/Graph/Graph.swift [hunk 2]
```diff
@@ -305,13 +311,44 @@ class Graph: Graphing {
...
+        }
+
+        dependencies.append(contentsOf:
+            resourceBundleDependencies(path: path, name: target.name)
+                .map { .product(target: $0.target.name, productName: $0.target.productNameWithExtension) })
+
+        return Set(dependencies).sorted()
+    }
+
+    func allDependencyReferences(for project: Project, system: Systeming) throws -> [DependencyReference] {
```
Comment (pepibumur) [inline Sources/TuistGenerator/Graph/Graph.swift]: We really need to get rid of the dependency injection for the `System` class 😬

**Comment by pepibumur** [comment]
The PR is looking good @kwridan 👏

**Comment by kwridan** [comment]
Thanks for taking a look @pepibumur!

File: Sources/TuistGenerator/Graph/Graph.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistGenerator/Graph/Graph.swift]: 👍 added a comment to https://github.com/tuist/tuist/pull/515/files#diff-12f3cd68bf1720345626c27c16f99056R108

File: Sources/TuistGenerator/Generator/ProjectFileElements.swift [hunk 3]
```diff
@@ -153,6 +147,20 @@ class ProjectFileElements {
                              isReference: $0.isReference)
         })
 
+        // Local Packages
```
Comment (kwridan) [inline Sources/TuistGenerator/Generator/ProjectFileElements.swift]: I need to think about this further ... 

The inconsistency here is that all other dependencies are resolved via `allDependencyReferences` whereas this one isn't included there yet.

**Pull request merged**